### PR TITLE
add k0s support

### PR DESCRIPTION
**Pull request by brunnels**
#1120 

- Support for [k0s](https://github.com/k0sproject/k0s)
- Replace `democratic-csi` with `openebs-localpath`
- Add support for templating into a private Github repo
- Rename bootstrap vars `bootstrap_kube_vip_addr` -> `bootstrap_kubeapi_addr` and `bootstrap_local_path_provisioner_path` -> `bootstrap_local_storage_path`
- Update Cilium to be `1.15.0` compatible

### CODE REVIEW

**Comment by onedr0p** [comment]
Thanks for taking the time on adding this! Few comments on this PR:

1. I think we should have this variable at the top of the config file and renamed from

	```yaml
	# set true to modify templates to work in k0s environment
	k0s:
	  enabled: false
	```
	
	to

	```yaml
    # (Advanced) Distribution can either be 'k3s' or 'k0s'
    # Note that changing this to k0s will ignore `system_upgrade_controller` in the addons.
	bootstrap_distribution: k3s
	```

2. I would like to have either `openebs-local-storage` OR `democratic-csi`. The latter I've tested and know works on k3s. I can do testing on my end if you really want to move forward with `openebs`. It doesn't matter too much to me which solution we choose as long as it has the option to specify the data path on the host.
3. `system-upgrade-controller` will need to be disabled even if the user tries to enable it in the addons.
4. `bootstrap_private_github_repo` only seems to have checks that verifies the private repo exists, however Flux needs to be modified and a secret needs to be added to support this. I originally left this out because I wanted people to keep it public to get better support. After they had it working and running they could read the README to make it private.
5. k3s needs to have certain metrics [dropped in a few places in kube-prometheus-stack](https://github.com/onedr0p/flux-cluster-template/blob/a7eadd34b11c8e92686acb8021acf72845a7328a/bootstrap/templates/addons/kube-prometheus-stack/app/helmvalues.yaml.j2#L32-L38), I don't think that should matter too much for k0s. My main question is does k0s work well with monitoring the `kubelet`, `kubeApiServer`, `kubeControllerManager` and `kubeEtcd` components out of the box or is there something that needs to be enabled or changed in the `kube-prometheus-stack` Helm values.
6. Lastly, you mentioned the nodes password field in the config. If you want to remove this option and clean up the places it is referenced I am fine with that since SSH done with a private SSH key and the nodes should have passwordless sudo.

File: bootstrap/templates/kubernetes/apps/kube-system/cilium/app/helmvalues.yaml.j2 [hunk 1]
```diff
@@ -16,7 +16,11 @@ data:
       id: 1
     containerRuntime:
       integration: containerd
+{% if k0s.enabled | default(false) %}
+      socketPath: /var/run/k0s/containerd.sock
+{% else %}
       socketPath: /var/run/k3s/containerd/containerd.sock
+{% endif %}
```
Comment (onedr0p) [inline bootstrap/templates/kubernetes/apps/kube-system/cilium/app/helmvalues.yaml.j2:23]:
```suggestion
      {% if k0s.enabled | default(false) %}
      socketPath: /var/run/k0s/containerd.sock
      {% else %}
      socketPath: /var/run/k3s/containerd/containerd.sock
      {% endif %}
```

Minor nit as it keeps styling the same

File: bootstrap/templates/kubernetes/apps/kube-system/kustomization.yaml.j2 [hunk 2]
```diff
@@ -5,5 +5,7 @@ kind: Kustomization
 resources:
   - ./namespace.yaml
   - ./cilium/ks.yaml
+{% if bootstrap_distribution == "k3s" %}
   - ./coredns/ks.yaml
+{% endif %}
```
Comment (onedr0p) [inline bootstrap/templates/kubernetes/apps/kube-system/kustomization.yaml.j2:10]:
```suggestion
  {% if bootstrap_distribution == "k3s" %}
  - ./coredns/ks.yaml
  {% endif %}
```

Minor nit as it keeps styling the same

File: bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2 [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  values:
+    localprovisioner:
+      hostpathClass:
+        enabled: true
+        isDefaultClass: false
+{% if bootstrap_distribution == 'k3s' %}
+        basePath: "{{ bootstrap_local_storage_path | default('/var/lib/rancher/k3s/local-hostpath', true) }}"
+{% else %}
+        basePath: "{{ bootstrap_local_storage_path | default('/var/openebs/local', true) }}"
+{% endif %}
```
Comment (onedr0p) [inline bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2:35]:
```suggestion
    localprovisioner:
      hostpathClass:
        enabled: true
        name: openebs-hostpath
        isDefaultClass: false
        basePath: "{{ bootstrap_local_storage_path | default('/var/openebs/local', true) }}"
```

1. We can keep the `basePath` the same and then include a task in the k3s ansible nuke playbook to delete this directory.
2. I also added the storageclass name here just to be declarative on this.

File: bootstrap/templates/kubernetes/apps/storage/openebs/app/kustomization.yaml.j2 [hunk 4]
```diff
@@ -2,4 +2,4 @@
 apiVersion: kustomize.config.k8s.io/v1beta1
 kind: Kustomization
 resources:
-  - ./helmrelease.yaml
+  - ./helmrelease.yaml
```
Comment (onedr0p) [inline bootstrap/templates/kubernetes/apps/storage/openebs/app/kustomization.yaml.j2:5]:
```suggestion
  - ./helmrelease.yaml

```

Missing a newline

File: bootstrap/templates/kubernetes/apps/storage/openebs/ks.yaml.j2 [hunk 5]
```diff
@@ -2,19 +2,19 @@
...
     labels:
       app.kubernetes.io/name: *app
-  path: ./kubernetes/apps/storage/local-path-provisioner/app
+  path: ./kubernetes/apps/kube-system/openebs/app
   prune: true
   sourceRef:
     kind: GitRepository
     name: home-kubernetes
-  wait: false
+  wait: true
```
Comment (onedr0p) [inline bootstrap/templates/kubernetes/apps/storage/openebs/ks.yaml.j2:17]:
```suggestion
  wait: false
```

We don't need `wait: true` here since applications can use a HelmRelease `dependsOn` instead of a Flux Kustomization `dependsOn`

File: bootstrap/vars/config.sample.yaml [hunk 6]
```diff
@@ -63,8 +69,11 @@ bootstrap_cluster_cidr: 10.42.0.0/16
 # (Advanced) For ipv6 use format 10.43.0.0/16,fd78:c889:47fb:e0::/112
 bootstrap_service_cidr: 10.43.0.0/16
 
-# (Advanced) Path for democratic-csi local-hostpath to use for storage on ALL nodes
-bootstrap_local_path_provisioner_path: /var/lib/rancher/k3s/local-hostpath
+# (Advanced) Path for openebs openebs-hostpath to use for storage on ALL nodes
+# for k3s distribution
+#bootstrap_local_storage_path: /var/lib/rancher/k3s/local-hostpath
+# for k0s distribution
+#bootstrap_local_storage_path: /var/openebs/local
```
Comment (onedr0p) [inline bootstrap/vars/config.sample.yaml:76]:
```suggestion
# (Advanced) Path for openebs openebs-hostpath to use for storage on ALL nodes
bootstrap_local_storage_path: /var/openebs/local
```

We can now default this.

File: bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2 (hunk 3 above)
Comment (onedr0p) [reply bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2:35]: Feel free to add number 1, or I can checkout the branch and add it. I suppose the `nuke` for k0s would also need it as well.

File: .taskfiles/K0sTasks.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+      - k0sctl apply --config k0s-config.yaml
+      - task: kubeconfig
+    preconditions:
+      - { msg: "k0s-config.yaml not found",      sh: "test -f {{.ROOT_DIR}}/k0s-config.yaml" }
+
+  reset:
+    desc: Resets the k0s cluster
+    cmd: k0sctl reset --config k0s-config.yaml
+    preconditions:
+      - { msg: "k0s-config.yaml not found",      sh: "test -f {{.ROOT_DIR}}/k0s-config.yaml" }
```
Comment (onedr0p) [inline .taskfiles/K0sTasks.yaml:25]:
```suggestion
    preconditions:
      - { msg: "k0s-config.yaml not found", sh: "test -f {{.ROOT_DIR}}/k0s-config.yaml" }

  apply:
    desc: Apply k0s cluster k0s-config.yaml
    cmds:
      - k0sctl apply --config k0s-config.yaml
      - task: kubeconfig
    preconditions:
      - { msg: "k0s-config.yaml not found", sh: "test -f {{.ROOT_DIR}}/k0s-config.yaml" }

  reset:
    desc: Resets the k0s cluster
    cmd: k0sctl reset --config k0s-config.yaml
    preconditions:
      - { msg: "k0s-config.yaml not found", sh: "test -f {{.ROOT_DIR}}/k0s-config.yaml" }
```

Minor nit on spacing in this playbook, since there's only 1 precondition we don't need to pad out the objects in the map.

File: bootstrap/tasks/optional/coredns.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
+---
```
Comment (onedr0p) [inline bootstrap/tasks/optional/coredns.yaml:1]: Should we put this in addons dir instead? I know it's not a toggle-able component but it would keep things in one place and I am fine with that.

File: bootstrap/tasks/validation/vars.yaml [hunk 9]
```diff
@@ -25,7 +26,6 @@
     - bootstrap_ipv6_enabled
     - bootstrap_k8s_gateway_addr
     - bootstrap_kube_vip_addr
-    - bootstrap_local_path_provisioner_path
```
Comment (onedr0p) [inline bootstrap/tasks/validation/vars.yaml:28]: We can add `bootstrap_local_storage_path` back here.

File: bootstrap/templates/kubernetes/flux/repositories/helm/openebs.yaml.j2 [hunk 10]
```diff
@@ -2,8 +2,8 @@
 apiVersion: source.toolkit.fluxcd.io/v1beta2
 kind: HelmRepository
 metadata:
-  name: democratic-csi
+  name: openebs
```
Comment (onedr0p) [inline bootstrap/templates/kubernetes/flux/repositories/helm/openebs.yaml.j2:5]: You will also need to update the `kustomization.yaml` file in the same directory.

File: bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2 (hunk 3 above)
Comment (brunnels) [reply bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2]: I made your changes.  Just need to add the cleanup to the nuke task.

File: bootstrap/vars/config.sample.yaml (hunk 6 above)
Comment (brunnels) [reply bootstrap/vars/config.sample.yaml]: Right, I added it to the required values validator and removed the default value in the template as well

File: bootstrap/templates/kubernetes/apps/storage/openebs/ks.yaml.j2 [hunk 11]
```diff
@@ -2,19 +2,19 @@
...
-  name: &app local-path-provisioner
+  name: &app openebs
   namespace: flux-system
 spec:
   targetNamespace: storage
   commonMetadata:
     labels:
       app.kubernetes.io/name: *app
-  path: ./kubernetes/apps/storage/local-path-provisioner/app
+  path: ./kubernetes/apps/kube-system/openebs/app
```
[the remark this answers is not in the document]
Comment (brunnels) [reply bootstrap/templates/kubernetes/apps/storage/openebs/ks.yaml.j2:12]: Good catch!

File: bootstrap/tasks/optional/coredns.yaml (hunk 8 above)
Comment (brunnels) [reply bootstrap/tasks/optional/coredns.yaml:1]: Could do since there's only one.

File: Taskfile.yaml [hunk 12]
```diff
@@ -47,6 +47,7 @@ tasks:
     desc: Configure repository from Ansible vars
     prompt: Any conflicting config in the root kubernetes and ansible directories will be overwritten... continue?
     dir: "{{.BOOTSTRAP_DIR}}"
-    cmd: ansible-playbook configure.yaml
+    cmds:
+      - ../.venv/bin/ansible-playbook configure.yaml
```
Comment (mrueg) [inline Taskfile.yaml:51]: is this change intended?

Comment (onedr0p) [reply Taskfile.yaml:51]: Ah yeah. I didn't catch this. The env _should_ contain everything needed for `ansible-playbook` to be found (at least it works on my machine). Also`cmd` was introduced in a recent taskfile update. Maybe @brunnels had some other ideas behind this change.

Comment (brunnels) [reply Taskfile.yaml:51]: If you just run `task configure` and haven't activated the env you will get an error message.  This fixes that.  Same thing is already being done here https://github.com/onedr0p/flux-cluster-template/blob/main/.taskfiles/AnsibleTasks.yaml#L27

Comment (onedr0p) [reply Taskfile.yaml:51]: Good point! I would make this line be `cmd: ../.venv/bin/ansible-playbook configure.yaml` then

Comment (brunnels) [reply Taskfile.yaml:51]: code will be updated in my next push

Comment (onedr0p) [reply Taskfile.yaml:51]: You could also remove `dir: "{{.BOOTSTRAP_DIR}}"` and update the cmd to match (might be best)

Comment (onedr0p) [reply Taskfile.yaml:51]: ```cmd: ansible-playbook {{.BOOTSTRAP_DIR}}/configure.yaml```

Comment (brunnels) [reply Taskfile.yaml:51]: Sure, will be in next push

Comment (mrueg) [reply Taskfile.yaml:51]: Have you considered setting a path variable instead?

Comment (onedr0p) [reply Taskfile.yaml:51]: I wish that would work, but it's and issue upstream with Taskfile: https://github.com/go-task/task/issues/482

Comment (onedr0p) [reply Taskfile.yaml:51]: In fact I am doing that, but it doesn't work and should be commented out with a `# TODO`:
https://github.com/onedr0p/flux-cluster-template/blob/main/.taskfiles/AnsibleTasks.yaml#L6

File: .taskfiles/K0sTasks.yaml [hunk 13]
```diff
@@ -0,0 +1,25 @@
+---
+version: "3"
+
+tasks:
+
+  kubeconfig:
+    desc: Gets k0s cluster kubeconfig
+    cmds:
+      - k0sctl kubeconfig -c k0s-config.yaml > kubeconfig
```
Comment (onedr0p) [inline .taskfiles/K0sTasks.yaml:9]:
```suggestion
      - k0sctl kubeconfig --cconfig k0s-config.yaml > kubeconfig
```
Nit on consistency with the tasks below this one

File: bootstrap/templates/k0s-config.yaml.j2 [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+{% if bootstrap_nodes.worker | default([]) | length > 0 %}
+  {% for item in bootstrap_nodes.worker %}
+  - role: {{ item.role | default('worker') }}
+    ssh:
+      address: {{ item.address }}
+      user: {{ item.username }}
+  {% endfor %}
+{% endif %}
+  k0s:
+    version: v1.28.4+k0s.0
```
Comment (mrueg) [inline bootstrap/templates/k0s-config.yaml.j2:33]: can we use renovate bot to update this?

File: bootstrap/tasks/optional/main.yaml [hunk 15]
```diff
@@ -0,0 +1,7 @@
+---
+
+- name: Process optional opendns
```
Comment (mrueg) [inline bootstrap/tasks/optional/main.yaml:3]: This should be coredns instead of opendns?

File: bootstrap/templates/k0s-config.yaml.j2 (hunk 14 above)
Comment (onedr0p) [reply bootstrap/templates/k0s-config.yaml.j2:33]: Yep, we should be able to add in 

```yaml
# renovate: datasource=github-releases depName=k0sproject/k0s
version: v1.28.4+k0s.0
```

File: bootstrap/configure.yaml [hunk 16]
```diff
@@ -17,23 +17,30 @@
...
 
     - name: Template Sops configuration
       ansible.builtin.include_tasks: tasks/sops/main.yaml
 
+    - name: Template k0s configuration
+      when: bootstrap_distribution == "k0s"
+      ansible.builtin.include_tasks: tasks/k0s/main.yaml
+
     - name: Template Ansible configuration
       ansible.builtin.include_tasks: tasks/ansible/main.yaml
```
Comment (onedr0p) [inline bootstrap/configure.yaml:37]: I wonder if we should somehow make it so that the k3s config isn't templated out when someone chooses k0s because right now I think if you use k0s, the k3s Ansible will be there.

File: bootstrap/templates/k0s-config.yaml.j2 [hunk 17]
```diff
@@ -0,0 +1,118 @@
+#jinja2: trim_blocks: True, lstrip_blocks: True
```
Comment (onedr0p) [inline bootstrap/templates/k0s-config.yaml.j2:1]: Across this file I would keep the YAML consistency and indent arrays. e.g. instead of

```yaml
  hosts:
  {% for item in bootstrap_nodes.master %}
  {% if (item.role is not defined) and (bootstrap_nodes.master | length > 1) %}
  - role: controller+worker
```

use

```yaml
  hosts:
  {% for item in bootstrap_nodes.master %}
  {% if (item.role is not defined) and (bootstrap_nodes.master | length > 1) %}
    - role: controller+worker
```

File: bootstrap/templates/k0s-config.yaml.j2 [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+                  algorithm: maglev
+                  mode: {{ bootstrap_cilium_loadbalancer_mode | default('dsr', true) }}
+                localRedirectPolicy: true
+                operator:
+                  replicas: 1
+                  rollOutPods: true
+                rollOutCiliumPods: true
+                securityContext:
+                  privileged: true
+                tunnel: disabled
```
Comment (onedr0p) [inline bootstrap/templates/k0s-config.yaml.j2:118]:
```suggestion
                routingMode: native
```

This change was made awhile and is being removed in 1.15.0, so we can make this change now.

File: bootstrap/templates/k0s-config.yaml.j2 [hunk 19]
```diff
@@ -0,0 +1,118 @@
...
+          {% endfor %}
+        extensions:
+          helm:
+            repositories:
+            - name: cilium
+              url: https://helm.cilium.io
+            charts:
+            - name: cilium
+              chartname: cilium/cilium
+              version: "1.14.5"
```
Comment (onedr0p) [inline bootstrap/templates/k0s-config.yaml.j2:68]:
```suggestion
              # renovate: datasource=github-releases depName=cilium/cilium
              version: "1.14.5"
```

File: bootstrap/templates/k0s-config.yaml.j2 (hunk 14 above)
Comment (onedr0p) [inline bootstrap/templates/k0s-config.yaml.j2:33]:
```suggestion
    # renovate: datasource=github-releases depName=k0sproject/k0s
    version: "v1.28.4+k0s.0"
```

Quotes matter on this value.

File: bootstrap/templates/k0s-config.yaml.j2 [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+  {% endfor %}
+{% if bootstrap_nodes.worker | default([]) | length > 0 %}
+  {% for item in bootstrap_nodes.worker %}
+  - role: {{ item.role | default('worker') }}
+    ssh:
+      address: {{ item.address }}
+      user: {{ item.username }}
+  {% endfor %}
+{% endif %}
+  k0s:
```
Comment (onedr0p) [inline bootstrap/templates/k0s-config.yaml.j2:32]:
```suggestion
  telemetry:
    enabled: false
  k0s:
```

Sorry Mirantis.

File: bootstrap/templates/k0s-config.yaml.j2 (hunk 14 above)
Comment (onedr0p) [reply bootstrap/templates/k0s-config.yaml.j2:33]: Also renovate change needed:

```json
    // custom versioning
    {
      "description": "Use custom versioning for k0s/k3s",
      "matchDatasources": ["github-releases"],
      "versioning": "regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(?<compatibility>\\+k?s)(?<build>\\d+)$",
      "matchPackagePatterns": ["k0s", "k3s"]
    },
```

File: bootstrap/vars/config.sample.yaml [hunk 21]
```diff
@@ -63,23 +71,23 @@ bootstrap_cluster_cidr: 10.42.0.0/16
...
-bootstrap_local_path_provisioner_path: /var/lib/rancher/k3s/local-hostpath
+# (Advanced) Path for openebs openebs-hostpath to use for storage on ALL nodes
+bootstrap_local_storage_path: /var/openebs/local
 
 # Node information
 bootstrap_nodes:
   # Use only 1, 3 or more odd master nodes, recommended is 3
   master:
-    # - name:     # name of the master node (must match [a-z0-9-]+)
+    # - name:     # hostname of the master node (must match [a-z0-9-\.]+)
```
Comment (onedr0p) [inline bootstrap/vars/config.sample.yaml:81]:
```suggestion
    # - name:     # name or hostname of the master node (must match [a-z0-9-\.]+)
```

File: bootstrap/vars/config.sample.yaml [hunk 22]
```diff
@@ -63,23 +71,23 @@ bootstrap_cluster_cidr: 10.42.0.0/16
...
     #   address:  # ip address of the master node
     #   username: # ssh username of the master node
-    #   password: # password of ssh username for the master node
-    #   # external_address: # Only use when SSH is not reachable on the IP specified in 'address' field
+    #   role:     # (Optional) k0s only: can be controller (default) or controller+worker
+    #   # external_address: # (Optional) k3s only: Only use when SSH is not reachable on the IP specified in 'address' field
     # ...
   worker: # set to [] or omit if no workers are needed
-    # - name:     # name of the worker node (must match [a-z0-9-]+)
+    # - name:     # hostname of the master node (must match [a-z0-9-\.]+)
```
Comment (onedr0p) [inline bootstrap/vars/config.sample.yaml:88]:
```suggestion
    # - name:     # name or hostname of the master node (must match [a-z0-9-\.]+)
```

File: bootstrap/templates/k0s-config.yaml.j2 [hunk 23]
```diff
@@ -0,0 +1,118 @@
...
+          kubeProxy:
+            disabled: true
+          {%  if bootstrap_nodes.master | length > 1 %}
+          nodeLocalLoadBalancing:
+            enabled: true
+            type: EnvoyProxy
+          {% endif %}
+          provider: custom
+        telemetry:
+          enabled: false
```
Comment (onedr0p) [inline bootstrap/templates/k0s-config.yaml.j2:118]: Looking at the docs this key appears to be under the first spec, not `.spec.hosts.config`?

https://docs.k0sproject.io/v1.28.4+k0s.0/configuration/#spectelemetry

File: bootstrap/configure.yaml (hunk 16 above)
Comment (onedr0p) [reply bootstrap/configure.yaml:37]: https://github.com/onedr0p/flux-cluster-template/issues/1126

File: bootstrap/templates/k0s-config.yaml.j2 (hunk 23 above)
Comment (onedr0p) [reply bootstrap/templates/k0s-config.yaml.j2]: Updated.

File: bootstrap/tasks/addons/main.yaml [hunk 24]
```diff
@@ -21,7 +25,7 @@
   ansible.builtin.include_tasks: kube_prometheus_stack.yaml
 
 - name: Process addon system-upgrade-controller
-  when: system_upgrade_controller.enabled | default(false)
+  when: (bootstrap_distribution == "k3s") and (system_upgrade_controller.enabled | default(false))
```
Comment (onedr0p) [inline bootstrap/tasks/addons/main.yaml:28]:
```suggestion
  when:
    - bootstrap_distribution == "k3s"
    - system_upgrade_controller.enabled | default(false)
```

File: bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2 (hunk 3 above)
Comment (onedr0p) [reply bootstrap/templates/kubernetes/apps/storage/openebs/app/helmrelease.yaml.j2]: I added simple enough logic around the nuke playbook that should work with k0s now.

**Review by onedr0p — APPROVED** [review]
